Tagged templates ES3 & 5

ivogabe opened the pull request:
I've implemented basic tagged templates emit when targeting ES3 or ES5.

```typescript
foo `A ${1} B ${2} C`;
```
is compiled to
```javascript
foo(["A ", " B ", " C"], 1, 2);
```

See also #1590. I haven't updated the tests yet.

Arnavion commented:
Perhaps a bug with template strings even before your change, but:

```javascript
function foo(a: string[], b: number) { }

foo`A${ 1 }B${ 2, 3 }C`;
```

gives "error TS2346: Supplied parameters do not match any signature of call target."

Changing foo to take 3 number parameters `function foo(a: string[], b: number, c: number) { }` instead suppresses the type error, but the original is the one that's correct, of course.

---

Now, with your change, this emits

```javascript
foo(["A", "B", "C"], 1, 2, 3);
```

whereas it should emit `1, (2, 3)`

Arnavion commented:
Opened #1600 for the typechecker issue.

ivogabe commented:
Thanks! Fixed

DanielRosenwasser commented:
You should have a function like `emitDownlevelTaggedTemplate` which actually *shouldn't* call `emitTemplateExpression` IMO.

So I'd rather have the raw property on - if we're doing it, we should do it more correctly, and t's only a matter of time until someone actually wants that. I think that if we're supporting downlevel, let's just do it here. However, let's wait for input from @mhegazy and @jonathandturner.

Arnavion commented:
Using a single `__template` at the top of the file gets nasty with nested template strings.

```javascript
var x = foo`12\n3${ bar`45\n6${ 789 }` }`;
```

```javascript
var x = foo((__template = ["12\n3", ""], __template.raw = ["12\\n3", ""], __template), bar((__template = ["45\n6", ""], __template.raw = ["45\\n6", ""], __template), 789));
```

It does seem to be well-defined per the ES spec. __template is assigned 123... first and then overwritten with 456... before calling bar(), but foo is still called with 123... because that was what __template evaluated to at the time.

Maybe create a new temporary for each tagged template string, just like destructuring creates a new temporary for each destructuring assignment.

```javascript
var a = ["45\n6", ""]; a.raw = ["45\\n6", ""];
var b = bar(a, 789);
var c = ["12\n3", ""]; c.raw = ["12\\n3", ""];
var x = foo(c, b);
```

Arnavion commented:
For comparison, 6to5 uses a helper function:

```javascript
var _taggedTemplateLiteral = function (strings, raw) {
  return Object.freeze(Object.defineProperties(strings, {
    raw: {
      value: Object.freeze(raw)
    }
  }));
};

var x = foo(_taggedTemplateLiteral(["12\n3", ""], ["12\\n3", ""]), bar(_taggedTemplateLiteral(["45\n6", ""], ["45\\n6", ""]), 789));
```

ivogabe commented:
I wouldn't prefer a function for this, it doesn't really reduce the complexity very much since the code with the variable isn't very complicated. I don't see an advantage (yet) of using multiple vars. I could see the advantage of having one variable per function (instead of per file), since uglifyjs for example can mangle those names very well.

Arnavion commented:
The advantage of multiple vars or the utility function is that it avoids using the same variable for multiple invocations, like in the nested template string example I gave.

Granted, I haven't been able to think of a way to break it, since ES guarantees that
- parameters are evaluated left-to-right, and
- function bodies bind their arguments at the time they're called

Since the strings and strings.raw expressions can only contain strings, not code, they can't have any other side effects on the shared variable apart from the assignment. Any code which does have side effects (such as a template expression containing a tagged template string) will appear after them in the parameter list and so won't affect them.

It just looks ugly and unreadable, and you have to resort to language-lawyering like above to demonstrate its correctness. It's also arguably not the kind of JS a human would write, which is one of TS's principles.

So my personal order of preference is:
1. One var per tagged template string. Used by Traceur.
2. One function call per tagged template string. Used by 6to5 (single function, reused for all strings) and jstransform (individual IIFE per string).
3. A single variable shared by multiple template strings, either file-level or function-level as you later suggested.

Arnavion commented:
@DanielRosenwasser 

>Function level variables are "annoying" to emit - you need to keep track of them with a flag - probably in NodeCheckFlags. The problem is where would you keep track of how many tagged template invocations you use?

Destructuring assignment already emits a local variable per assignment. Looking at the code, it just tries increasing variable names until it finds one that isn't available in the current scope and all parent scopes. Temporaries are kept track of in a tempVariables array that is updated according to the temporaries in the current scope as emit traverses the AST.

>Template string arrays are supposed to be immutable.

You could use Object.freeze for ES5 and not for ES3, or you could just not use it altogether as a compromise. If you go with the utility function route you can let the user override it with their own implementation of the function (ala __extends) that does use Object.freeze() if they so desire. That said...

>Referential equality breaks if we do not properly intern the array, whereas the spec mandates this. I believe each template array should be the same so long as the constituent raw portions are equal.

This is interesting. I didn't know about this, but you're right. (Template strings are registered in a registry, and if the strings and raw strings of a particular template are identical to one already in the registry, the registered string's template object is used for the first argument to the tag function. See [12.3.7](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-tagged-templates) and [12.2.8.2.2](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-runtime-semantics-gettemplateobject))

FWIW it does not seem to be followed by FF at the moment.

```javascript
function foo(arr) { return arr; }
foo`123` === foo`123`; // false
```

Perhaps it would be fine for TS's emit to also not bother.

Or, as you said, you could use Object.freeze() to atleast maintain immutability, and not support this downlevel emit for ES3.

DanielRosenwasser commented:
Spoke with @mhegazy about this. We'd like to get this in for 1.5 if possible.

The approach we're favoring is actually one similar to destructuring - one variable for every invocation. This is to favor a reader's perspective when trying to reason about the code, so it may be more ideal. If you need any pointers in pursuing this direction, just let us know.

ivogabe commented:
Ok, will try to do that. Just to be sure, where will the variable statement be emitted? I think we want before the statement which contains a tagged template. Thus if the tagged template is a part of another tagged template and another statement, it will be written above that statement.

The only problem or inconsistency I see with this approach is the case when a tagged template is used within the condition of a for (or while) loop. The idea is that a template array won't be reused but will be reassigned every time it's needed. But in a for loop, how will that work? Since if you'd emit it like I described it above, it will emit:
```
var a = ["Lorem", "Ipsum"];
a.raw = ["Lorem", "Ipsum"];
for (var i = 0; foo(a, i); i++) {}
// Ugly alternative:
var a;
for (var i = 0; a = ["Lorem", "Ipsum"],
a.raw = ["Lorem", "Ipsum"], foo(a, i); i++) {}
```
Is this an edge-case that doesn't matter?

Arnavion commented:
Just do the same as destructuring:

```javascript
var x, y;
for (; { x, y } = { x: 5, y: 6 };) {
}
```

```javascript
var x, y;
for (; (_a = { x: 5, y: 6 }, x = _a.x, y = _a.y, _a);) {
}
var _a;
```

ivogabe commented:
Ok, I've implemented the variable approach, except for the loops. Current codegen:

```typescript
// Typescript
foo `${foo `A${1}B`}`;

// Javascript
var _a = ["", ""];
_a.raw = ["", ""];
var _b = ["A", "B"];
_b.raw = ["A", "B"];
foo(_a, foo(_b, 1));
```

Have I implemented the functionality in the right functions and files?

Also when you have a comment before a statement with a tagged template, the temporary variable will be emitted before the comment:
```typescript
// Typescript

// Comment 
foo `${foo `A${1}B`}`;

// Javascript

var _a = ["", ""];
_a.raw = ["", ""];
var _b = ["A", "B"];
_b.raw = ["A", "B"];
// Comment
foo(_a, foo(_b, 1));
```
Can this be solved easily?

DanielRosenwasser commented on src/compiler/binder.ts:
```diff
@@ -371,9 +372,25 @@ module ts {
         }
+        
+        function bindTaggedTemplateExpression(node: TaggedTemplateExpression) {
+            if (file.languageVersion >= ScriptTarget.ES6) return;
```
Newline/curlies around the `if` body.

DanielRosenwasser commented on src/compiler/types.ts:
```diff
@@ -703,6 +704,7 @@ module ts {
 
     export interface Statement extends Node, ModuleElement {
         _statementBrand: any;
+        downlevelTaggedTemplates?: TaggedTemplateExpression[]; // Initialized in binder.ts
```
This really does't seem appropriate; I believe we manage to get emit for destructuring without tacking a property on any other node.

Arnavion commented:
For lambda expressions returning the tagged template string, the temporaries are generated outside the lambda instead of inside. Regular functions are fine:

```javascript
var bar = () => foo`A${ 1 }B`;
(() => foo`C${ 2 }D`)();
var baz = function () { return foo`E${ 3 }F`; };
(function () { return foo`G${ 4 }H`; })();
```

```javascript
var _a = ["A", "B"];
_a.raw = ["A", "B"];
var bar = function () { return foo(_a, 1); };
var _b = ["C", "D"];
_b.raw = ["C", "D"];
(function () { return foo(_b, 2); })();
var baz = function () {
    var _a = ["E", "F"];
    _a.raw = ["E", "F"];
    return foo(_a, 3);
};
(function () {
    var _a = ["G", "H"];
    _a.raw = ["G", "H"];
    return foo(_a, 4);
})();
```

_a and _b for the first two could be emitted inside the anonymous function instead, similar to the last two.

Arnavion commented:
This crashes the compiler:

```javascript
function foo(...args) { }

class A {
	x = foo`A${ 1 }B`;
}
```

```
C:\Stuff\Sources\TypeScript\built\local\tsc.js:6576
            statement.downlevelTaggedTemplates.push(node);
                     ^
TypeError: Cannot read property 'downlevelTaggedTemplates' of undefined
    at bindTaggedTemplateExpression (C:\Stuff\Sources\TypeScript\built\local\tsc.js:6576:22)
    at bind (C:\Stuff\Sources\TypeScript\built\local\tsc.js:6691:21)
    at child (C:\Stuff\Sources\TypeScript\built\local\tsc.js:3185:24)
    at Object.forEachChild (C:\Stuff\Sources\TypeScript\built\local\tsc.js:3217:179)
    at bindChildren (C:\Stuff\Sources\TypeScript\built\local\tsc.js:6462:16)
    at bindDeclaration (C:\Stuff\Sources\TypeScript\built\local\tsc.js:6506:13)
    at bind (C:\Stuff\Sources\TypeScript\built\local\tsc.js:6608:21)
    at children (C:\Stuff\Sources\TypeScript\built\local\tsc.js:3194:34)
    at Object.forEachChild (C:\Stuff\Sources\TypeScript\built\local\tsc.js:3328:139)
    at bindChildren (C:\Stuff\Sources\TypeScript\built\local\tsc.js:6462:16)
```

(You can use these as tests even if you change the code per DanielRosenwasser's comments.)

ivogabe commented:
@DanielRosenwasser Thanks for the feedback. I think the difficulty is that for destructuring, the variable is emitted in the current node (if it is a variable declaration) or before it (for example in a for loop). The variable is added to an array and when it's possible to emit these variable declarations it will emit them. The variable declarations are thus emitted after the destructuring node, but that doesn't matter since the variable will be hoisted.

For these templates we want the variables before the tagged template, since only the declaration, not the initialization will be hoisted. An alternative would be to emit
```javascript
for (var i = 0; foo((_a = ["Lorem", "Ipsum"], _a.raw = ["Lorem", "Ipsum"], _a), i); i++) {}
var _a;
```
But we decided we didn't want that. So I think it's necessary to determine where those temporary variables should be emitted before the emit actually starts. Or am I missing something?

@Arnavion It searches for a statement that is the parent of the tagged template, but as you pointed out that isn't working always, since not all tagged templates have to have a statement as parent, or it might point to the wrong parent.

DanielRosenwasser commented:
@ivogabe assignment expressions permit binding patterns on the left-hand side as well, so destructurings are actually *also* an expression-level construct.

Currently the emit for those temps occurs at the end of the enclosing scope - this is because it is permissible given JavaScript's hoisting semantics.

See `emitTempDeclarations` for where this occurs, but all you'll have to concern yourself with is `ensureIdentifier` and `recordTempDeclaration` in the emitter. Just extract those two functions out of `emitDestructuring` so that you can use them and the rest should follow.

ivogabe commented:
@DanielRosenwasser If I look at the generated code for destructuring, it will be possible with that approach to emit tagged templates like this:
TypeScript:
```typescript
console.log(foo `a${1}b`);
```
JavaScript:
```javascript
console.log(foo((_a = ['a', 'b'], _a.raw = ['a', 'b'], _a), 1));
var _a;
```
or
```javascript
console.log((_a = ['a', 'b'], _a.raw = ['a', 'b'], foo(_a, 1)));
var _a;
```

But the proposed emit was
```javascript
var _a = ['a', 'b'];
_a.raw = ['a', 'b'];
console.log(foo(_a, 1));
```
I think the last emit isn't possible with that approach, because it will need to know when such var statements need to be emitted. That should be determined before the emit really starts, because the initialization of the variables need to happen before the tagged template. Or is there a better way to do this?

DanielRosenwasser commented:
@ivogabe the proposed emit is a hassle in retrospect.

Between 

```Javascript
console.log(foo((_a = ['a', 'b'], _a.raw = ['a', 'b'], _a), 1));
var _a;
```
and
```Javascript
console.log((_a = ['a', 'b'], _a.raw = ['a', 'b'], foo(_a, 1)));
var _a;
```

The first one is sort of nice in that the function call doesn't get shifted around, but I think the second is a little easier on the eyes; personally, it's easier to parse out the string components and the arguments.

So I'd say that for now, do the second version.

ivogabe commented:
@DanielRosenwasser Ok, will try to implement that this week. This makes life easier indeed!

Arnavion commented:
:+1:

The outer parens are sometimes emitted when they're not needed (when the tagged template is a return expression or element access expression) but it doesn't seem worth detecting such cases. (These are also  issues for regular template strings anyway.)

DanielRosenwasser commented on src/compiler/emitter.ts:
```diff
@@ -2050,6 +2050,67 @@ module ts {
                 return '"' + escapeString(node.text) + '"';
             }
+            
+            function emitDownlevelRawTemplateLiteral(node: LiteralExpression, isLast: boolean) {
```
Forget `isLast`, just check the `kind`.

DanielRosenwasser commented on src/compiler/emitter.ts:
```diff
@@ -2050,6 +2050,67 @@ module ts {
+                // Last template piece ends with "`", others with "${"
+                text = text.substring(1, text.length - (isLast ? 1 : 2));
+                
+                write('"' + escapeString(text) + '"');
```
Newline normalization still takes place here as per the spec:

> The TRV of *LineTerminatorSequence* :: \<CR> is the code unit value 0x000A.
> The TRV of *LineTerminatorSequence* :: \<CR>\<LF> is the sequence consisting of the code unit value 0x000A.

So you'll need to replace `\r\n`s and `\r`s with `\n`s.

I'd do something like

```TypeScript
text = text.replace(/\r\n?/g, "\n");
text = escapeString(text);

write('"' + text + '"');
```

I think we should have tests for a lot of this.

DanielRosenwasser commented on src/compiler/emitter.ts:
```diff
@@ -2050,6 +2050,67 @@ module ts {
+                emit(tempVariable);
+                write(".raw = [");
+                if (node.template.kind === SyntaxKind.NoSubstitutionTemplateLiteral) {
+                    emitDownlevelRawTemplateLiteral(<LiteralExpression> node.template, true);
```
Remove the spaces following the type assertions.

DanielRosenwasser commented on src/compiler/emitter.ts:
```diff
@@ -2050,6 +2050,67 @@ module ts {
+                        write(", ");
+                        var needsParens = templateSpan.expression.kind === SyntaxKind.BinaryExpression
+                            && (<BinaryExpression> templateSpan.expression).operator === SyntaxKind.CommaToken;
+                        emitParenthesized(templateSpan.expression, needsParens);
```
Can we rename this `emitParenthesizedIf`?

DanielRosenwasser commented:
From what I can offhandedly see, this looks good! First pull in and merge with the latest `master` branch. Then let's see what the baselines look like. We should have some decent coverage already, but it's good to be paranoid about these things.

DanielRosenwasser commented:
Hey @ivogabe, no rush, but wondering what the status is on this PR.

ivogabe commented:
I was a bit busy with some exams this week. Anyway, I tried to accept new baselines, but then I get over 3000 new files in the git diff (of which most are not related to tagged templates). I'm working on a Mac, if that does matter. Will try it on Windows next week.

DanielRosenwasser commented:
No worries, very odd though - that might be a newline issue, but we'll have to figure that out. If you try it on a Windows machine, WinDiff has a pretty decent folder comparison tool (or you can splurge and get BeyondCompare which is awesome).

ivogabe commented:
I've updated the tests. I still got 3000 changed files on Windows, so OS didn't matter. The files that changed that begin with `taggedTemplate` or `template` look pretty good in my opinion. I don't know why other files changed as well.

What tests should be added? I think the raw property should be tested properly, the other parts are already tested a lot.

DanielRosenwasser commented:
@ivogabe, this is very odd. Our Jakefile supposed to delete the `projectOutput` directory in the local baselines. I'm not sure how you got a handle on it after running `jake runtests`. Once you do a `jake baseline-accept`, everything is usually fair game to add in the `baseline/references` directory.

Do you think you could amend your last commit and unstage the `projectOutput` directory?

Should be something like

1. `git rm -r tests/baselines/reference/projectOutput`
2. `jake runtests`
3. `jake baseline-accept`
4. Ensure that `tests/baselines/reference/projectOutput` is not present, and if it is, delete it.
5. `git add tests/baselines/reference/*`
6. `git commit --amend`
7. `git push --force origin taggedTemplates`

ivogabe commented:
Thanks for your help, now I got only 32 changed files. That looks better!

What do you think of the new baselines?

Arnavion commented:
https://github.com/ivogabe/TypeScript/blob/8e16e1d010cca171c1da63bc285c6bccc50049c3/tests/baselines/reference/templateStringInTaggedTemplate.js   is wrong. Instead of

```javascript
(_a = ["I ", " THE TEMPLATE PORTION"], _a.raw = ["I ", " THE TEMPLATE PORTION"], "I AM THE " + "TAG" + " " + " PORTION"(_a, "AM"));
var _a;
```

it needs to be

```javascript
(_a = ["I ", " THE TEMPLATE PORTION"], _a.raw = ["I ", " THE TEMPLATE PORTION"], ("I AM THE " + "TAG" + " " + " PORTION")(_a, "AM"));
var _a;
```

DanielRosenwasser commented:
Hey @ivogabe, as @Arnavion mentioned, the emit is incorrect for the `"I AM THE" + "TAG" + PORTION"` example. You can use the `needsParenthesisForPropertyAccess` function. If you do use this, please rename it to `needsParenthesesForPropertyAccessOrInvocation`.

ivogabe commented:
@Arnavion Thanks for finding that issue!

I've fixed it and updated the baselines. It also wrapped object literals with parens. I think that's also better this way.

@DanielRosenwasser I think I also found out why I got problems with updating the baselines. `jake runtests` executes mocha. Since the tests don't pass, mocha aborts. Because of that, the jake process aborts too, and the command to remove the temporary directory isn't executed. By running `jake runtests`, `jake baseline-accept`, `jake runtests`, `jake baseline-accept`, the directory will be deleted, because the tests pass the second time. Could that be the problem?

DanielRosenwasser commented:
I'm not sure if that could exactly be the problem - tests not passing is supposed to happen prior to a `baseline-accept`. I am familiar with a problem with mocha on Windows which has caused a few issues for us, requiring some sort of patch. Perhaps @danquirk or @RyanCavanaugh might know better.

Also, you'll need to merge in from master so that we can get an appropriate idea of the final PR. Once you merge in from master, you'll likely need to update the baselines again (sorry!)

Arnavion commented:
I think @ivogabe meant that running runtests and baseline-accept once leaves the projectOutput/ directory behind, because runtests aborts on the test failures before it can call deleteTemporaryProjectOutput(). Running runtests a second time after having accepted the baselines succeeds, and deletes the projectOutput/ directory.

runtests should definitely exit with a fail() if tests fail as it does now, otherwise test failures won't fail the build in Travis, etc. I guess baseline-accept should also call deleteTemporaryProjectOutput() on success to fix this issue.

ivogabe commented:
I've merged the master again & updated the tests. Good to see there's a lot happening in the master :+1: 

I think @Arnavion describes how I meant it. When running runtests, the process is aborted before it can delete the temporary directory, because the full jake process is aborted since mocha failed.

DanielRosenwasser commented:
@mhegazy and I will discuss a solution, and it's being tracked on #2050.

DanielRosenwasser commented:
So I'm surprised that I myself didn't get any tests on this (or rather, that I did, but they were for *untagged* templates), but can you please create 6 tests (3 in ES6 mode, 3 unspecified) out of the following?

--------

`taggedTemplateStringsWithMultilineTemplate.ts` / `taggedTemplateStringsWithMultilineTemplateES6.ts`
```TypeScript
function f(...args: any[]): void {
}

f `


`;
```

--------

`taggedTemplateStringsWithWhitespaceEscapes.ts` / `taggedTemplateStringsWithWhitespaceEscapesES6.ts`
```TypeScript
function f(...args: any[]) {
}

f `\t\n\v\f\r`;
```

--------

`taggedTemplateStringsWithUnicodeEscapes.ts` / `taggedTemplateStringsWithUnicodeEscapesES6.ts`
```TypeScript
function f(...args: any[]) {
}

f `'\u{1f4a9}'${ " should be converted to " }'\uD83D\uDCA9'";
```

--------

To specify that a test should be in ES6 mode, you simply add `//@target: es6` to the top of the file

DanielRosenwasser commented:
*(As a heads up, I've just updated my comment for the first test and specified the line endings)*

ivogabe commented:
In the last test, I think you ended the template with the wrong quote, or was that intentional? That test also throws error TS1125: Hexadecimal digit expected at `\u{1f4a9}` (after correcting the quote).

Arnavion commented:
Yes, it's expected to fail at the moment, because TS doesn't support ES6 unicode escapes yet. See #2047

DanielRosenwasser commented:
Yes my mistake, I had the wrong quote, though @Arnavion's right about the intentional failure on unicode escapes. Maybe it would be best to add a test like:

`taggedTemplateStringsHexadecimalEscapes.ts` / `taggedTemplateStringsHexadecimalEscapesES6.ts`
```TypeScript
function f(...args: any[]) {
}

f `\x0D${ "Interrupted CRLF" }\x0A`;
```

ivogabe commented:
I've added the tests

DanielRosenwasser commented on src/compiler/emitter.ts:
```diff
@@ -2138,6 +2138,72 @@ module ts {
+                }
+                else {
+                    emitDownlevelRawTemplateLiteral((<TemplateExpression>node.template).head);
+                    forEach((<TemplateExpression>node.template).templateSpans, (child, index) => {
```
`index` actually isn't used here.

DanielRosenwasser commented:
:100: great job @ivogabe! I'm really happy with the results here, so here's my :+1:

Let's get another sign off and we'll pull in the changes.

CyrusNajmabadi commented on src/compiler/emitter.ts:
```diff
@@ -2138,6 +2138,72 @@ module ts {
+                var isLast = node.kind === SyntaxKind.NoSubstitutionTemplateLiteral || node.kind === SyntaxKind.TemplateTail;
+                text = text.substring(1, text.length - (isLast ? 1 : 2));
+                
+                // Newline normalization
```
Mention the portion of the es6 spec that specifies this.

DanielRosenwasser replied:
```TypeScript
// ES6 Spec 11.8.6.1 - Static Semantics of TV's and TRV's
// <CR><LF> and <CR> LineTerminatorSequences are normalized to <LF> for both TV and TRV.
```

CyrusNajmabadi commented on src/compiler/emitter.ts:
```diff
@@ -2138,6 +2138,72 @@ module ts {
             }
+            
+            function emitDownlevelRawTemplateLiteral(node: LiteralExpression) {
+                var text = getSourceTextOfNodeFromSourceFile(currentSourceFile, node);
```
I'm torn.  On one hand, this brings us closer to what the user already wrote.  ON the other, it means you need to do a lot more processing.  Why not just use node.text instead?  It will already have stripped off the } and ${  bits, and will already have changes newlines appropriately.

CyrusNajmabadi replied:
Oh.  Is this because it's supposed to be 'raw'?  If so, can you comment here the difference and how this shoudl behave on different inputs?  Thanks!

CyrusNajmabadi commented on src/compiler/emitter.ts:
```diff
@@ -2138,6 +2138,72 @@ module ts {
+            
+            function emitDownlevelTaggedTemplate(node: TaggedTemplateExpression) {
+                var tempVariable = createTempVariable(node);
+                recordTempDeclaration(tempVariable);
```
dan added a new createAndRecordTempVariable.

DanielRosenwasser replied:
Good catch

CyrusNajmabadi commented:
Looks great!  I just had a few nits to take care of (mostly about commenting functions to make their purpose and behavior clear).

JsonFreeman commented on src/compiler/emitter.ts:
```diff
@@ -2138,6 +2138,72 @@ module ts {
+                
+                // Newline normalization
+                text = text.replace(/\r\n?/g, "\n");
+                text = escapeString(text);
```
Aren't these steps for the cooked strings?

JsonFreeman replied:
Oh maybe these are the steps that uncook the string

JsonFreeman commented on src/compiler/emitter.ts:
```diff
@@ -2138,6 +2138,72 @@ module ts {
+                    forEach((<TemplateExpression>node.template).templateSpans, (child, index) => {
+                        write(", ");
+                        emitDownlevelRawTemplateLiteral(child.literal);
+                    });
```
Seems like the same thing as above, except you call emitDownlevelRawTemplateLiteral instead of emit. I'd make a helper that takes a function, and call it twice, passing in emitDownlevelRawTemplateLiteral or emit as appropriate.

JsonFreeman commented on src/compiler/emitter.ts:
```diff
@@ -2138,6 +2138,72 @@ module ts {
+                }
+                write("], ");
+                
+                emitParenthesizedIf(node.tag, needsParenthesisForPropertyAccessOrInvocation(node.tag));
```
Why do you need parentheses? I don't think you need to add them. The reason is that you want something that can be the left hand side of a call. But MemberExpression and CallExpression are the two nonterminals that can be the LHS of a tagged template, and those are precisely the things you need on the LHS of a call.

JsonFreeman replied:
Unless there is some subtlety with `new` that I'm missing, but I don't think there is.

Arnavion replied:
As nonsensical as it is, it's because of https://github.com/Microsoft/TypeScript/pull/1589#issuecomment-73362303

JsonFreeman replied:
Yeah looking at the grammar, you can't have a NewExpression as the tag, so you don't need to parenthesize the tag.

JsonFreeman replied:
Oh gosh, you're right! The template literal gets demoted to an AdditiveExpression! Disregard what I said.

JsonFreeman replied:
But I also thought we parenthesize templates as part of emitting *them*. Wouldn't it already be parenthesized?

Arnavion replied:
For example, 6to5 also adds parens: http://bit.ly/1zWUBIa

Arnavion replied:
Yes, that makes sense. A CallExpression should be standalone and have parens around it if necessary, without needing the code to add them.

JsonFreeman replied:
But do we do the right thing in this case?
```ts
`hello${"goodbye"}hello`[0]
```

JsonFreeman replied:
My point is that rather than each parent node needing to know that its child could be a template, and therefore needs to parenthesize that template, the template should parenthesize itself accordingly, because *it* is the one that lowers its precedence when it gets emitted.

JsonFreeman replied:
To clarify, when I say template in this context, I mean an *untagged* template.

Arnavion replied:
Yes, I was agreeing with you that node.tag should include the parens, instead of requiring the code that uses it to add them.

For your example, the current PR also emits it correctly `("hello" + "goodbye" + "hello")[0];` but yes, your suggested implementation would be better.

JsonFreeman replied:
Yes, I think that's a better approach. Although I do think this PR can go in as is, and we make that change separately.

JsonFreeman commented on src/compiler/emitter.ts:
```diff
@@ -2138,6 +2138,72 @@ module ts {
+                    forEach((<TemplateExpression>node.template).templateSpans, templateSpan => {
+                        write(", ");
+                        var needsParens = templateSpan.expression.kind === SyntaxKind.BinaryExpression
+                            && (<BinaryExpression>templateSpan.expression).operator === SyntaxKind.CommaToken;
```
Yes, comma is the only expression lower than assignment

JsonFreeman commented on src/compiler/emitter.ts:
```diff
@@ -2408,7 +2475,7 @@ module ts {
             }
 
-            function needsParenthesisForPropertyAccess(node: Expression) {
+            function needsParenthesisForPropertyAccessOrInvocation(node: Expression) {
```
Leave as is

JsonFreeman replied:
Nvm, leave as is.

JsonFreeman commented:
:+1:

ivogabe commented:
Thanks for the feedback everyone! I've added some new commits, can you re-review it?

@CyrusNajmabadi That was indeed for the raw strings. I've added a comment to explain it. Is that clear enough?

DanielRosenwasser commented:
Odd, the baselines seem to have changed for a few files. I swear, we're almost there! :wink:

ivogabe commented:
That's strange, on my machine they pass. I've run the tests again and they don't error. Any idea how that's happening?

DanielRosenwasser commented:
The tests pass on my system too, so try pulling from master and pushing again. Travis might just need to be triggered.

ivogabe commented:
There were some baselines that changed after merging the master, should be fixed now. Does Travis pull commits from both the master as the branch that is tested?

DanielRosenwasser commented:
Yeah, I believe what happens is that Travis will run tests on a given branch and then on a hypothetically merged branch. Just waiting on @mhegazy to take a look at this change.

DanielRosenwasser commented:
Due to tests introduced by #2143, can you pull in from master and update the baselines?

mhegazy commented:
:+1: 

thanks @ivogabe

ivogabe commented:
I've merged the master again. It looks like taggedTemplateStringsWithUnicodeEscapes is wrong, the `\u` before `{1f4a9}` is missing?

DanielRosenwasser commented:
No that's fine, it's meant to be broken until #2047 is fixed. Actually, I'm more concerned that we don't preserve string escapes in the output, but I'll take care of it there where we'll have easier means.

Well I think that's it - really big thanks for this feature! Our team, and I think the community, will definitely be happy to see this, and you've done a great job!
